### DISCUSSION TITLE
A database statement execution error occurred: NOT NULL constraint failed: item.type

### DISCUSSION THREAD

**Discussion by DieBe** [abraunegg/onedrive#2936] [Q&A] [answered]
First of all, thanks for the great project with an outstanding documentation!

I've been using your onedrive client successfully for quite some time now. First from the official Ubuntu repository, then, when I realized how old that version was, I switched to building it from source using the commands provided in the documentation with the dmd compiler. I also started using onedriveGUI to be able to get some information on what the client is doing.
I don't know when exactly (and I don't think the exact point in time is too relevant) I, by accident, disabled the sync_list I was previously using, causing the client to download far more than I wanted it to.
When I re-enabled the sync_list obviously a resync was required. After that, I realized that obviously the files I never wanted to download were still there (even though not synced).
Because I didn't want to run in any risk of deleting this data from my onedrive by deleting it locally while ```onedrive --monitor``` is running, I decided to remove the whole onedrive folder and resync again. I did this with a simple ```rm -rf ~/OneDrive``` and ```onedrive --sync --resync```.  To be extra sure I also reinstalled the onedrive client, using ```make uninstall``` and the build commands from the documentation. I also deleted the my ```~.config/onedrive/``` folder, backing up my ```config``` and ```sync_list``` and putting them back before the initial sync.
When I finally did sync the new installation, this error showed up:
```
A database statement execution error occurred: NOT NULL constraint failed: item.type

Please restart the application with --resync to potentially fix any local database issues.
```

Restarting with ```--resync``` doesn't seem to change anything.

I tried to resolve this myself, but nothing worked so far, I also tried rebuilding it multiple times.  I also tried deleting the database manually, but that also doesn't change anything.
The only way I get it to sync is by running it with ```sudo```,  but I don't like the idea of running processes with sudo as a workaround.

Is there any way of resolving this? I would have thought that reinstalling the client should resolve it (as the database should be regenerated)...

I'm using the latest Kubuntu 24.10 (I know, it's your favourite :) ). Below is my configuration:
```
--------------- Application Runtime Configuration ---------------
Application version                          = onedrive v2.5.2-13-g967327c
Compiled with                                = DMD 2109
User Application Config path                 = /home/benjamin/.config/onedrive
System Application Config path               = /etc/onedrive
Applicable Application 'config' location     = /home/benjamin/.config/onedrive/config
Configuration file found in config location  = true - using 'config' file values to override application defaults
Applicable 'sync_list' location              = /home/benjamin/.config/onedrive/sync_list
Applicable 'items.sqlite3' location          = /home/benjamin/.config/onedrive/items.sqlite3
Config option 'drive_id'                     = 
Config option 'sync_dir'                     = ~/OneDrive
Config option 'enable_logging'               = false
Config option 'log_dir'                      = /var/log/onedrive
Config option 'disable_notifications'        = false
Config option 'skip_dir'                     = .Trash-1000|Application\ Data|Dokumente
Config option 'skip_dir_strict_match'        = true
Config option 'skip_file'                    = ~*|.~*|*.tmp|*.swp|*.partial
Config option 'skip_dotfiles'                = false
Config option 'skip_symlinks'                = false
Config option 'monitor_interval'             = 300
Config option 'monitor_log_frequency'        = 12
Config option 'monitor_fullscan_frequency'   = 12
Config option 'read_only_auth_scope'         = false
Config option 'dry_run'                      = false
Config option 'upload_only'                  = false
Config option 'download_only'                = false
Config option 'local_first'                  = false
Config option 'check_nosync'                 = false
Config option 'check_nomount'                = false
Config option 'resync'                       = true
Config option 'resync_auth'                  = false
Config option 'cleanup_local_files'          = false
Config option 'classify_as_big_delete'       = 1000
Config option 'disable_upload_validation'    = false
Config option 'disable_download_validation'  = false
Config option 'bypass_data_preservation'     = false
Config option 'no_remote_delete'             = false
Config option 'remove_source_files'          = false
Config option 'sync_dir_permissions'         = 700
Config option 'sync_file_permissions'        = 600
Config option 'space_reservation'            = 52428800
Config option 'application_id'               = d50ca740-c83f-4d1b-b616-12c519384f0c
Config option 'azure_ad_endpoint'            = 
Config option 'azure_tenant_id'              = 
Config option 'user_agent'                   = ISV|abraunegg|OneDrive Client for Linux/v2.5.2-13-g967327c
Config option 'force_http_11'                = false
Config option 'debug_https'                  = false
Config option 'rate_limit'                   = 0
Config option 'operation_timeout'            = 3600
Config option 'dns_timeout'                  = 60
Config option 'connect_timeout'              = 10
Config option 'data_timeout'                 = 60
Config option 'ip_protocol_version'          = 0
Config option 'threads'                      = 14
Compile time option --enable-notifications   = false

Selective sync 'sync_list' configured        = true
sync_list config option 'sync_root_files'    = false
sync_list contents:



!/Bilder/*
!/Desktop/*
!lib/
!cache/

/temp/*
/Dokumente/*
/Backup/*
/thunderbird/*
/Code/*

Config option 'sync_business_shared_items'   = false

Config option 'webhook_enabled'              = false

--------------------DEVELOPER_OPTIONS----------------------------
Config option 'force_children_scan'          = false

-----------------------------------------------------------------

```

Thanks in advance!

**Comment by abraunegg** [comment, marked as answer]
@DieBe 
> The only way I get it to sync is by running it with sudo, but I don't like the idea of running processes with sudo as a workaround.

Firstly this is a *major* problem .. you should not be running `sudo onedrive <application options>` to run the client.

This means that everything is running as 'root' , where the config file sits, where the database file sits, the permissions on your files etc ... 

What you really need to do is *fix* that issue first .. and that is something I cannot help with at all. You should be able to run the application *without* running sudo - so that all files, config, database etc are operating as your normal user.

Please prioritise fixing this issue on your platform before working on the next issue.

> When I finally did sync the new installation, this error showed up:
> `A database statement execution error occurred: NOT NULL constraint failed: item.type`
>

This potentially sounds like a bug, but *also* could be a side impact of running things via sudo.

### Actions for you
1. Ensure that you have removed all 'onedrive' package versions from your system given this statement: First from the official Ubuntu repository .... you must remove this as the binary file if still present will be conflicting
2. Install the 'latest' master code , which at the time of writing this is `onedrive v2.5.2-25-g2267a23`. Ensure that you can run the client *without* sudo
3. If you still encounter this issue / problem, raise a new bug report and ensure you provide a debug log created using this process: https://github.com/abraunegg/onedrive/wiki/Generate-debug-log-for-support

As the OpenSuSE Build Service repository does not support Ubuntu 24.10 that is why there are no pre-built packages available.

I would be therefor raising a bug with Ubuntu to make the correct client version available by making the Ubuntu teams provide an updated package version.

**Reply by DieBe** [reply]
Thanks for the quick reply!
I never intended to actually use it with ```sudo```, but I thought it might be good to know.

Concerning the actions:
1. I used ```sudo find . -name "onedrive"``` running at the root directory to find every directory or file called onedrive and deleted all of those (and their contents). The only remaining directory is ```/var/lib/sddm/.config/onedrive```, which a) is empty and b) I don't want to mess up sddm.
2. I previously installed it in ```/usr/local/src/``` (which required sudo), now I chose ```~/bin/``` to be able to install it without sudo. The installation worked, but when running ```onedrive --sync --resync --reauth```, after the reauth, the same error occurs.
3. Based on this I'll create a bug report.

**Comment by abraunegg** [comment]
You should always *install* using sudo - `sudo make install` ...  so that the application installs to the *correct* system locations.

Just do not *run* the application via sudo.

This is all called out in the documentation.

**Reply by DieBe** [reply]
Maybe I should clarify this:
Originally I had to use ```sudo``` in front of every part of the installation commands, now I only used it with ```sudo make install```, not with ```make```, ```./configure``` or ```make clean```.
